numpy/core/src/_simd/_simd_inc.h.src

```diff
@@ -0,0 +1,257 @@
+    simd_data data;
+    // argument object
+    PyObject *obj;
+} simd_arg;
+
+typedef struct
+{
+    PyObject_HEAD
+    // vector type id
+    simd_data_type type;
+    // vector data, aligned for safe casting
+    npyv_lanetype_u8 NPY_DECL_ALIGNED(NPY_SIMD_WIDTH) data[NPY_SIMD_WIDTH];
+} simd_vector;
+
+/************************************
+ ** Declarations (inc_data)
+ ************************************/
+/**
+ * returns data info of certain dtype,
+ * NOTE: not protected against overflow
+ */
+static const simd_data_info *
+simd_data_getinfo(simd_data_type dtype);
+// returns 1 if dtype is a scalar, otherwise '0'
+static int
+simd_data_is_scalar(simd_data_type dtype);
+// returns 1 if dtype is a sequence, otherwise '0'
+static int
+simd_data_is_sequence(simd_data_type dtype);
+// returns 1 if dtype is a vector, otherwise '0'
+static int
+simd_data_is_vector(simd_data_type dtype);
+// returns 1 if dtype is a multi-vector x2, otherwise '0'
+static int
+simd_data_is_vectorx2(simd_data_type dtype);
+// returns 1 if dtype is a multi-vector x3, otherwise '0'
+static int
+simd_data_is_vectorx3(simd_data_type dtype);
+/**
+ * converts dtype to equivalent vector type,
+ * e.g. u8 -> vu8
+ */
+static simd_data_type
+simd_data_cvt2vector(simd_data_type dtype);
+/**
+ * converts dtype to equivalent scalar type,
+ * e.g. vu8 -> u8
+ */
+static simd_data_type
+simd_data_cvt2scalar(simd_data_type dtype);
+
+/************************************
+ ** Declarations (inc_convert)
+ ************************************/
+/**
+ * convert PyObject to simd_data, raise Python exception on
+ * failure.
+ * NOTE: simd_obj2data_clear() must be called after get done
+ * from returned data.
+ * example:
+ ** simd_data d = simd_obj2data(obj, simd_data_qu8);
+ ** ...
+ ** simd_obj2data_clear(d, simd_data_qu8);
+ */
+static simd_data
+simd_obj2data(PyObject *obj, simd_data_type dtype);
+/**
+ * convert simd_data to PyObject, raise Python exception on
+ * failure and return NULL.
+ */
+static PyObject *
+simd_data2obj(simd_data data, simd_data_type dtype);
+
+static int
+simd_data2seq(PyObject *seq, simd_data data, simd_data_type dtype);
+/**
+ * free memory allocations, that caused by simd_obj2data,
+ * mainly to free sequences pointer.
+ */
+static void
+simd_obj2data_clear(simd_data data, simd_data_type dtype);
+
+/************************************
+ ** Declarations (inc_vector)
+ ************************************/
+/**
+ * convert simd_data to PyObject(simd_vector),
+ * raise Python exception on failure and returns NULL.
+ */
+static simd_vector *
+simd_data2vector(simd_data data, simd_data_type type);
+/**
+ * convert PyObject(simd_vector) to simd_data,
+ * raise Python exception on failure.
+ */
+static simd_data
+simd_vector2data(simd_vector *vec, simd_data_type type);
+/**
+ * init and register vector type(PyTypeObject) to PyModule,
+ * vector type can be reached through attribute 'vector_type'.
+ * return 0 on error, 1 on success.
```

Comment:
This isn't the usual convention for python C api functions - usually it's -1 on error, 0 on success.